blockly/src/starter/PathfinderStarter.java

```diff
@@ -0,0 +1,115 @@
+package starter;
+
+import contrib.entities.HeroFactory;
+import contrib.level.DevDungeonLoader;
+import contrib.systems.*;
+import contrib.systems.LevelEditorSystem;
+import core.Entity;
+import core.Game;
+import core.components.CameraComponent;
+import core.game.ECSManagment;
+import core.systems.LevelSystem;
+import core.utils.Tuple;
+import core.utils.components.path.SimpleIPath;
+import java.io.IOException;
+import java.util.logging.Level;
+import level.AiMazeLevel;
+import systems.PathfindingSystem;
+import utils.CheckPatternPainter;
+import utils.pathfinding.DFSPathFinding;
+
+/** This class starts the dungeon Ai level to visualize the DFS and BFS. */
+public class PathfinderStarter {
+  private static final boolean DRAW_CHECKER_PATTERN = true;
+
+  /**
+   * Setup and run the game. Also start the server that is listening to the requests from blockly
+   * frontend.
+   *
+   * @param args
+   * @throws IOException
+   */
+  public static void main(String[] args) throws IOException {
+    Game.initBaseLogger(Level.WARNING);
+
+    // start the game
+    configGame();
+    // Set up components and level
+    onSetup();
+
+    onLevelLoad();
+
+    // build and start game
+    Game.run();
+  }
+
+  private static void onSetup() {
+    Game.userOnSetup(
+        () -> {
+          DevDungeonLoader.addLevel(Tuple.of("dfs", AiMazeLevel.class));
+          createSystems();
+
+          createHero();
+
+          LevelSystem levelSystem = (LevelSystem) ECSManagment.systems().get(LevelSystem.class);
+          levelSystem.onEndTile(DevDungeonLoader::loadNextLevel);
+
+          DevDungeonLoader.loadLevel(0);
+        });
+  }
+
+  private static void onLevelLoad() {
+    Game.userOnLevelLoad(
+        (firstLoad) -> {
+          if (DRAW_CHECKER_PATTERN)
+            CheckPatternPainter.paintCheckerPattern(Game.currentLevel().layout());
+
+          PathfindingSystem pathfindingSystem =
+              (PathfindingSystem) ECSManagment.systems().get(PathfindingSystem.class);
```

Comment:
`ECSManagment.systems()` bricht die Kapselung. Das gehört endlich mal ersetzt durch sowas in der Art:

```java
  public static <T extends System> Optional<T> system(Class<? extends System> s) {
    return Optional.ofNullable((T) SYSTEMS.get(s));
  }
```

Noch besser wäre, wenn man die auszuführende Aktion in die Methode in `ECSManagment` reinreichen würde und nicht interne Elemente rausreichen würde (in meinem Vorschlag würde ja immer noch ein System rausgereicht).

Reply:
Hier ein Lösungsvorschlag mit Umkehr der Dependeny:

(1) obiger Code:
```java
          ECSManagment.system(
              PathfindingSystem.class,
              (pfs) -> {
                pfs.autoStep(true);
                pfs.updatePathfindingAlgorithm(
                    new DFSPathFinding(
                        Game.currentLevel().startTile().coordinate(),
                        Game.currentLevel().endTile().coordinate()));
              });
```

(2) In `ECSManagment`:

```java
  public static <T extends System> void system(Class<T> s, Consumer<T> c) {
    if (SYSTEMS.containsKey(s)) c.accept((T) SYSTEMS.get(s));
  }
```